src/counterfactual_mean_based/estimands.jl

```diff
@@ -245,22 +245,91 @@ unique_treatment_values(dataset, colnames) =
+"""
+If there is no dataset and the treatments_levels are a NamedTuple, then they are assumed correct.
+"""
+make_or_check_treatment_levels(treatments_levels::NamedTuple, dataset::Nothing) = treatments_levels
+
+"""
+If no dataset is provided, then a NamedTuple precising treatment levels is expected
+"""
+make_or_check_treatment_levels(treatments, dataset::Nothing) = 
+    throw(ArgumentError("No dataset from which to infer treatment levels was provided. Either provide a `dataset` or a NamedTuple `treatments` e.g. (T=[0, 1, 2],)"))
+
+"""
+If a list of treatments is provided as well as a dataset then the treatment_levels are infered from it.
+"""
+make_or_check_treatment_levels(treatments, dataset) = unique_treatment_values(dataset, treatments)
+
+"""
+If a NamedTuple of treatments_levels is provided as well as a dataset then the treatment_levels are checked from the dataset.
+"""
+function make_or_check_treatment_levels(treatments_levels::NamedTuple, dataset)
+    for (treatment, treatment_levels) in zip(keys(treatments_levels), values(treatments_levels))
+        dataset_treatment_levels = Set(skipmissing(Tables.getcolumn(dataset, treatment)))
+        missing_levels = setdiff(treatment_levels, dataset_treatment_levels)
+        length(missing_levels) == 0 || 
+            throw(ArgumentError(string("Not all levels provided for treatment ", treatment, " were found in the dataset: ", missing_levels)))
+    end
+    return treatments_levels
+end
+
+function _factorialEstimand(
+    constructor, 
+    treatments_settings::NamedTuple{names}, outcome; 
+    confounders=nothing,
+    outcome_extra_covariates=nothing,
+    freq_table=nothing,
+    positivity_constraint=nothing,
+    verbosity=1
+    ) where names
+    components = []
+    for combo ∈ Iterators.product(values(treatments_settings)...)
+        Ψ = constructor(
+            outcome=outcome,
+            treatment_values=NamedTuple{names}(get_treatment_setting(combo)),
+            treatment_confounders = confounders,
+            outcome_extra_covariates=outcome_extra_covariates
+        )
+        if satisfies_positivity(Ψ, freq_table; positivity_constraint=positivity_constraint)
+            push!(components, Ψ)
+        else
+            verbosity > 0 && @warn("Sub estimand", Ψ, " did not pass the positivity constraint, skipped.")
+        end
+    end
+    if length(components) == 0
+        throw(ArgumentError("No component passed the positivity constraint."))
+    end
+    return JointEstimand(components...)
+end
+
+"""
+    factorialEstimand(
+        constructor::Union{typeof(CM), typeof(ATE), typeof(IATE)},
+        treatments, outcome; 
+        confounders=nothing,
+        dataset=nothing, 
+        outcome_extra_covariates=(),
+        positivity_constraint=nothing,
+        freq_table=nothing,
+        verbosity=1
+    )
+
+Generates a factorial `JointEstimand` with components of type `constructor` (CM, ATE, IATE). 
```

Comment:
This now generates a joint estimand instead of a composed estimand. Is this so that we can call composed with a function of interest later?

Reply:
yes, I've added a PR description above.